### DISCUSSION TITLE
Unable to include a directory within an excluded directory

### DISCUSSION THREAD

**Discussion by benjamindaines** [borgbase/vorta#1901] [FAQ] [answered]
I am trying to exclude the .cache directory in my home directory, but I do want the yay directory within .cache to be backed up.  I've tried adding it to the source directories, and have tried using +/home/ben/.cache/yay* within the excludes, however it does not back up.  This seems to be the proper usage of the + according to borg documentation.  Suggestions?  Thanks.

**Comment by real-yfprojects** [comment, marked as answer]
Vorta passes your excludes to borg using the `--exclude-from` cmd option. Thus the [advanced pattern style](https://borgbackup.readthedocs.io/en/stable/usage/help.html) using `+`/`-`/`R`/... doesn't work. 
You can still achieve the desired behaviour:

```regex
re:^/home/ben/.cache/(?!yay)
```

**Reply by goebbe** [reply]
Another option would be the use of `--pattern` or `--patterns-from` - via the "Schedule" tab and then the "Shell Commands" section.
See here for examples: https://github.com/borgbase/vorta/issues/1991#issuecomment-2225916293

Edit: Note that these use "advanced borg features", that are not build into Vorta. Read the relevant docs before you uses these.

**Reply by benjamindaines** [reply]
> Vorta passes your excludes to borg using the `--exclude-from` cmd option. Thus the [advanced pattern style](https://borgbackup.readthedocs.io/en/stable/usage/help.html) using `+`/`-`/`R`/... doesn't work. You can still achieve the desired behaviour:
> 
> ```regexp
> re:^/home/ben/.cache/(?!yay)
> ```

I just saw your response now, thank you.  FINALLY have it doing what I want.  I suspected that the issue was that the excludes were being passed with --exclude-from.  However, why?  Would it not make more sense to have, especially the raw, tab behave as a normal exclude list (meaning includes can be added as well)? 

EDIT: this may not even work.  When I first plugged that line into the raw section, the file counts looked appropriate.  However, upon recalculating, the file count is reflecting the entire .cache directory being backed up.  I'm going to run the update and see what happens, but I am not hopeful as I was just a second ago. 

EDIT EDIT: Yup, it's taking the entire cache directory :\ so that did not work.  It seems like I'm going to have to relocate my yay cache..... not something a user should have to do.

**Reply by goebbe** [reply]
@benjamindaines Personally I use `--pattern-from` with a separated pattern file. The _include_ and _exclude_ work reliably for me. 
However, there are some things to consider: 
- when you use _include_ and _exclude_ the **order** does matter.
- excluding a folder, e.g. `.cache` will exclude the folder from _future_ backups - but it will not delete the .cache folder from  old backups. 
- look at the borg documentation on how-to remove `.cache` folders from old backup archives (using `borg recreate` together with your new --pattern or --pattern-from i.e. `include` and `exclude` restrictions). Test this before you _recreate_ all your old archives. ;-)    https://borgbackup.readthedocs.io/en/stable/usage/recreate.html
- Test using `--dry-run` and double check before you use `borg recreate`

Using `--pattern` the following pattern could be used
`--pattern=-home/Username/**/.cache`

Using `--pattern-from` and a pattern file, the following line excludes all the .cache files and .cache folders of Username: 
```
## exclude .cache files and folders:
- home/Username/**/.cache
```

Warning: `--pattern` and `--pattern-from` are **borg** features that are not integrated into Vorta. Consider these as "advanced" features that require some reading.

**Reply by benjamindaines** [reply]
Correct, order does matter. That's why it was such a frustrating experience, because I was reading the documentation and it should have worked if it were for --exclude-from 

Patterns may be the better solution, and I see they're added to the UI now. However, I have not experimented with them. When I adopted vorta and originally made this config, patterns were not included. 

I think vorta directing users to Borg documentation adds to this problem as well. Without indication that vorta excludes are strictly excludes, Borg documentation gives the impression that you can do includes and excludes. 

Just things to consider. 

Or, abstract the config further and allow the user to pick excluded locations via file picker and calculate the appropriate config in the dark. Still allow text based config, but have the primary method of exclusion be Fischer Price style.

**Reply by goebbe** [reply]
@benjamindaines 
Do the Vorta docs refer to borg at all? 
Also in the docs, the use of _include_ or the use of `--pattern` is nowhere recommended. And the GUI only mentions _exclude_. 

I believe that _include_ is just something that can be useful - and this is the reason why we have enhancement request and workarounds, posted by users. 
These use cases go beyond what Vorta has built in. If somebody use these (more advanced) borg features he/she should consult the borg documentation, first. 

I fully agree, that this should be stressed more prominently.

**Reply by benjamindaines** [reply]
It doesn't look like they do.... pretty sure they did at one point though.  Again, this discussion was started a year and a half ago, and I'll admit that I did not look before making that statement ;)
